PSA: Phase 2: Add support for PSA-based opaque PSKs

hanno-arm opened the pull request:
__Summary:__ This PR adds support for the use of opaque, PSA-based PSKs in PSK-only ciphersuites.

__Changes:__ The PR makes the following changes:
1. ~~It adds identifiers and implementations for the key derivation mechanisms for the TLS-1.2 PRF and TLS-1.2 PSK-to-MasterSecret expansion to the PSA crypto API resp. implementation. __These are being reviewed separately.__~~ This has been merged into `feature-psa` in the meantime so is no longer present in this PR.
1. It adds two public API functions `mbedtls_ssl_conf_psk_opaque()` and `mbedtls_ssl_set_hs_psk_opaque()` for the static resp. dynamic configuration of opaque (PSA key slot based) PSKs. They parallel the existing API `mbedtls_ssl_conf_psk()`, `mbedtls_ssl_set_hs_psk()`. Please see the internal design document in IOTSSL-2551 for more information. __Commits:__ e5ec04e
2. It adds an implementation of the new API. __Commits:__  9307c5f b01c1c0 a9ec6c6
3. It adapts `mbedtls_ssl_derive_keys()` to make use of the PSA-based PSK-to-MS key derivation mechanism when an opaque key has been configured. _Note:_ The master secret is exported and maintained in RAM as before. __Commits:__ f611d4e b5ceefb
4. It adapts client- and server-specific code. __Commits:__ 4f51cf3, 07f09fe, f3e8d27, 32b1291
5. Add `psk_slot` and `psk_list_slot` command line parameters to `ssl_client2` and `ssl_server2` example application to test the new API. __Commits:__ 6416475, 6180162
6. Add tests for opaque PSKs to `ssl-opt.sh`: 1b05947 ef5df3e

__References:__ See IOTSSL-2551 for the internal design document for the new API.

**Note:** will need rebasing on the submodule work before merging! The "DO NOT MERGE" label can be removed once it's done.

hanno-arm commented:
@AndrzejKurek Thank you for your review. I believe I have addressed your feedback. Could you please take another look?

AndrzejKurek commented on library/ssl_tls.c:
```diff
@@ -621,6 +643,14 @@ int mbedtls_ssl_derive_keys( mbedtls_ssl_context *ssl )
+    size_t const master_secret_len = 48;
+
+#if defined(MBEDTLS_SSL_EXTENDED_MASTER_SECRET)
+    unsigned char session_hash[48];
```
Is this 48 unrelated to the master secret length?

hanno-arm replied:
Yes it is, which enforces your point about naming those occurrences which are.

AndrzejKurek commented on library/ssl_tls.c:
```diff
@@ -700,68 +730,127 @@ int mbedtls_ssl_derive_keys( mbedtls_ssl_context *ssl )
-                    hash_len = 48;
-                }
+                if( md_type == MBEDTLS_MD_SHA384 )
+                    salt_len = 48;
```
Is this 48 unrelated to the master secret length?

hanno-arm replied:
Yes it is, which enforces your point about naming those occurrences which are.

mpg commented on library/ssl_tls.c:
```diff
@@ -7351,12 +7347,34 @@ int mbedtls_ssl_conf_psk( mbedtls_ssl_config *conf,
+        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+    }
+
+    ssl_conf_remove_psk( conf );
```
I think it would be slightly better to call this right at the top of the function. That would make behaviour more consistent: this function always wipes out previous PSKs, and a new one is configured iff it returns 0. It would also protect a bit against misuse by users not checking the return value, as any error would leave no PSK configured.

hanno-arm replied:
Agreed.

mpg commented on library/ssl_tls.c:
```diff
@@ -7351,12 +7347,34 @@ int mbedtls_ssl_conf_psk( mbedtls_ssl_config *conf,
+        return( MBEDTLS_ERR_SSL_BAD_INPUT_DATA );
+
+    /* Identity len will be encoded on two bytes */
+    if( ( psk_identity_len >> 16 ) != 0 ||
```
Suggestion for code improvement: if you agree about moving `conf_remove_psk()` to the top, then we might factor out the part about identity (the `NULL` check at the top, this part, allocation + copy below) that's currently a bit duplicated between this function and its opaque variant.

hanno-arm replied:
Done.

mpg commented on programs/ssl/ssl_client2.c:
```diff
@@ -135,9 +140,23 @@ int main( void )
     "    psk_identity=%%s     default: \"Client_identity\"\n"
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+#define USAGE_PSK_SLOT                          \
+    "    psk_slot=%%d         default: 0\n"     \
```
As discussed in the PR that adds something similar for ECDSA keys, I think it would be better to call the option `psk_opaque` and make it a boolean. That way when PSA changed the way keys are accessed, we only need to change the option's implementation, not its interface.

hanno-arm replied:
Yes! At the time of writing the PR, I didn't expect a slot allocation mechanism, nor thought of our intermediate approach of searching through the slot list ourselves. I will adapt the PR to follow that approach, and consequently rename `psk_slot` to `psk_opaque`.

mpg commented on programs/ssl/ssl_client2.c:
```diff
@@ -994,6 +1030,76 @@ int main( int argc, char *argv[] )
     mbedtls_debug_set_threshold( opt.debug_level );
 #endif
 
+#if defined(MBEDTLS_KEY_EXCHANGE__SOME__PSK_ENABLED)
```
Question: why did this block of code have to move?

mpg commented on programs/ssl/ssl_client2.c:
```diff
@@ -994,6 +1030,76 @@ int main( int argc, char *argv[] )
+
+        if( opt.force_ciphersuite[0] <= 0 )
+        {
+            mbedtls_printf( "opaque PSKs are only supported in conjunction with forcing TLS 1.2 and a PSK-only ciphersuite through the 'force_ciphersuite' option.\n" );
```
Minor: we may want to add a comment that the rest (TLS version and pure-PSK key exchange) will be checked later, because otherwise it looks locally like the test in the code doesn't match the error message.

hanno-arm replied:
I will add a comment.

mpg commented on programs/ssl/ssl_client2.c:
```diff
@@ -1466,6 +1546,47 @@ int main( int argc, char *argv[] )
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+    if( opt.psk_slot != 0 )
+    {
+        status = psa_crypto_init();
```
Note: this should be removed when rebasing on the PR that adds it near the beginning of `main()`.

mpg commented on library/ssl_tls.c:
```diff
@@ -607,6 +607,28 @@ static void ssl_calc_finished_tls_sha384( mbedtls_ssl_context *, unsigned char *
 
+#if defined(MBEDTLS_KEY_EXCHANGE__SOME__PSK_ENABLED) && \
+    defined(MBEDTLS_USE_PSA_CRYPTO)
+static int ssl_use_opaque_psk( mbedtls_ssl_context const *ssl )
```
Minor: I'd prefer the name `ssl_psk_is_opaque()`. I find "use" slightly weird for a boolean function.

hanno-arm replied:
The name `ssl_psk_is_opaque()` doesn't reflect very well that there's not a single PSK in the SSL context, but one has to consider both the static and the dynamic PSK configuration to see which PSK to actually use. I'd be inclined to stick with the present name, but if you insist I'll change it.

mpg replied:
I won't insist.

mpg commented on programs/ssl/ssl_server2.c:
```diff
@@ -220,9 +226,38 @@ int main( void )
+    "                          id1,psk1[,id2,psk2[,...]]\n"
+#if defined(MBEDTLS_USE_PSA_CRYPTO)
+#define USAGE_PSK_SLOT                          \
+    "    psk_slot=%%d         default: 0 (don't use key slots)\n"     \
```
As for the client, perhaps we shouldn't expose slots in command line options.

hanno-arm replied:
Yes, I will switch to the approach of searching for free slots automatically.

mpg commented on programs/ssl/ssl_server2.c:
```diff
@@ -220,9 +226,38 @@ int main( void )
+    "                          Note: This is to test integration of PSA-based opaque PSKs with\n"     \
+    "                          Mbed TLS only. Production systems are likely to configure Mbed TLS\n"  \
+    "                          with prepopulated key slots instead of importing raw key material.\n" \
+    "    psk_list_slot=%%d    default: 0 (don't use key slots)\n"     \
```
Please consider removing this option and selecting slots dynamically instead.

hanno-arm replied:
I don't want to remove this option to allow us testing opaque static PSKs but raw dynamic ones, and vice versa. But as for `psk_slot` I will turn it into a boolean option, if you're ok with that.

mpg replied:
Oh, good point about testing a mix. Please keep that option then, but just rename it and turn it to a boolean.

hanno-arm commented on include/mbedtls/ssl.h:
```diff
@@ -2057,68 +2057,146 @@ int mbedtls_ssl_conf_own_cert( mbedtls_ssl_config *conf,
- * \note           Currently clients can only register one pre-shared key.
- *                 In other words, the servers' identity hint is ignored.
+ * \warning        Currently, clients can only register a single pre-shared key.
+ *                 Calling this function or mbedtls_ssl_conf_opaque_psk() more
```
Thanks, fixing

hanno-arm commented:
@mpg @AndrzejKurek I added commits 80cea23, e3e6199, 5df9b04 to address @mpg's feedback. 

Further, I rebased this PR on top of the utility PR #2184, keeping the old branch [here](https://github.com/hanno-arm/mbedtls/tree/opaque_psk_implementation_backup151118).

Please re-review.

hanno-arm commented:
@mpg @AndrzejKurek Now that `feature-psa` contains the new PSK KDFs needed for this PR, I rebased this PR on top of `psa-integration-utilities`. Further, 15bb275 addresses an uninitialized variable bug caught by the compiler.

Please re-review.

gilles-peskine-arm commented:
`all.sh` passes locally on 15bb275d5b2d953956833af5db7e6ec070716b78 except for the Mbed Crypto exporter which is about to be removed anyway.

AndrzejKurek commented on library/ssl_tls.c:
```diff
@@ -7450,44 +7450,56 @@ static void ssl_conf_remove_psk( mbedtls_ssl_config *conf )
-
     /* Identity len will be encoded on two bytes */
-    if( ( psk_identity_len >> 16 ) != 0 ||
+    if( psk_identity               == NULL ||
```
This formatting is a bit confusing.

hanno-arm commented:
@mpg @AndrzejKurek I rebased this on top of the rebased `psa-integration-utilities`. Please re-review.
